web/api/v1/api.go

```diff
@@ -805,6 +808,93 @@ func rulesAlertsToAPIAlerts(rulesAlerts []*rules.Alert) []*Alert {
 	return apiAlerts
 }
 
+type metricsWithMetadata map[string]metadataSet
+
+type metadataSet map[metadata]struct{}
+
+func (ms metadataSet) MarshalJSON() ([]byte, error) {
```

Comment:
You shouldn't need any custom marshalling code, and this doesn't appear to be doing anything non-standard.

Reply:
Without it, our JSON encoding library complains about it with `unsupported map key type: v1.metadata`

Reply:
Then use a string as the key.

Reply:
Sorry if this not a great question, but wouldn't I still need to have custom marshalling code to define how the string should be output?

Reply:
No, if you think you need that then it is your data structure which needs improvement.

Reply:
Sooo… the `metadataSet` has to be the way it is to de-duplicate the metadata, see L847 and following below. There is no way it can be "improved". Go has no native sets, so we use a map with the metadata struct as a key and "nothing" (i.e. an empty struct) as a value. So far so good.

Obviously, we cannot just marshal that `metadataSet`, as the complex key is not compatible with JSON. Plus, we don't even want that representation. We want a JSON array.

Now there are two possibilities:
1. Take the whole, possible very large `metadataSet` and copy it into a newly created `[]metadata`. That would naturally marshal into exactly what we want.
2. Create a custom `MarshalJSON` to teach the `metadataSet` to marshal into what we want.

(1) is probably a bit less code but also less efficient. (2) is the most efficient solution. I'd intuitively prefer (2) but I don't feel strongly. @gotjosh implemented (2). I am inclined to follow that path. Not sure, though, if the code of `MarshalJSON` could be improved? I'm not an expert in writing those methods.

Reply:
I wrote "possible very large `metadataSet`". That's wrong. Most of them will actually only have one element. The concern is that there are possibly very many `metadataSet`s, and we have to convert every single one of them into a slice.

Reply:
My concern is that we should never have to hand-write JSON marshalling code, it introduces unnecessary risk and makes things more difficult to maintain. The only existing one is due to it's extreme performance criticality.

If you need to convert a map to a list then you should do so before it hits JSON-related code, not inside the JSON code.

Reply:
I'll amend it @brian-brazil.

Reply:
You can move this into metricMetadata.

Reply:
I actually like the state right now most, i.e. have an explicit `MarshalJSON` but it doesn't contain any low-level JSON code generation but only the conversion to a slice.

If we did this without a `MarshalJSON`, we had to create an entire new `metricsForMarshaling := map[string][]metadata{}` and populate it in two nested `for` loops. That's a lot of logic in `metricMetadata`, and it is only about marshaling. That's a layering violation, to use one of @brian-brazil's favorite words.

The current state of this PR gets the best of both worlds: marshaling is well contained, and no hand-written JSON code generation is happening.

Reply:
That wouldn't be a layering violation, as you're not reaching into the internals of the JSON-related code or making inappropriate assumptions about how other layers work. It's the same way all standard JSON marshalling works.

However this code does contain layering violation, as it's using jsoniter which means it's depending on a particular JSON marshaller which is a bit of a smell.

I still want to avoid any use of MarshalJSON where it's not absolutely necessary. For example here the function is returning a map of maps, yet somehow the API returns a map of arrays. This is unneeded magic.

> populate it in two nested for loops.

I don't see a problem with making things explicit.

Reply:
The problem here is that what we want is a map of sets. Neither Go nor JSON have that data structure. So Go uses a map of maps while JSON uses a map of arrays.

Brokering between those two different compromises is a marshaling question.

To layer it properly, we needed it in a different function anyway. It could be a separate function called `convertToMapOfSlicesForJSONMarshaling` (slightly exaggerated to make my point), or it could be the `MarshalJSON` method on the Go data structure. The latter makes way more sense to me. It also makes more sense to the author of this PR. You like it differently, fine. It's two to one at this point, and I don't think this is worth an even longer discussion. I'd like to ask you to accept the "majority opinion".

Usage of jsoniter is indeed a smell. It would be better if we could avoid it here. But it is also used throughout this file at various places. It doesn't tip the balance for me to not prefer a separate conversion over a `MarshalJSON` method.

Reply:
> To layer it properly, we needed it in a different function anyway.

Not really, it's only a few lines of code. But if a different function ends up containing those lines of code I can live with that.

> convertToMapOfSlicesForJSONMarshaling

Who said anything about JSON? The current endpoint functions don't known that JSON is involved, it could just as well be going to YAML or XML. That we wish to flatten a map into its values isn't related to the marshalling or JSON themselves.

> But it is also used throughout this file at various places.

Only for overall marshalling, not inside any of the functions for specific endpoints. There is currently a separation of concerns, which this PR would remove.

Reply:
As the name says, `MarshalJSON` is specific to JSON marshaling. And the method is not specific to an endpoint but to the data structure. We are essentially doing the same with `promql.Point`, just that this data structure is not from this package, so adding a method to it would be inappropriate.

Converting to a map of slices rather than a map of sets _is_ specific to JSON (as JSON has no sets).

If I had infinite time, I'd happily continue our little argument, but we are way beyond diminishing returns here. I suggest once more to merge this as is.

Reply:
> As the name says, MarshalJSON is specific to JSON marshaling.

That's my argument, you'd need to do the exact same thing if it were YAML or most other formats. Thus why this is a smell.

If it's inside the function endpoint then it'd work for everything.

> We are essentially doing the same with promql.Point, just that this data structure is not from this package, so adding a method to it would be inappropriate.

Not quite, and that's on an extremely important hot path (query/query_range) that I couldn't do a breaking change to when I was optimising that. That's one of the reasons I don't want any MarshallJSON, so that this doesn't end up needing more handcoded marshalling in the future. Nor do I want a bad pattern being introduced that other things then follow.

Using MarshallJSON is possibly less performant too, compared to letting jsoniter just handle all of this in the usual way.

>  I suggest once more to merge this as is.

I don't believe that code health concerns should be ignored because consensus hasn't been found. I think a strong argument is needed to introduce an inconsistent pattern which also has been known to be problematic in the past, and the arguments presented thus far do not approach that bar.

Reply:
> Who said anything about JSON? The current endpoint functions don't know that JSON is involved, it could just as well be going to YAML or XML.

Think this makes a great point. The current functions don't know anything about how the output should be rendered. The introduction of `MarshalJSON` breaks this pattern.

I'll amend it.

Reply:
The last commit contains the changes made. Can I get (hopefully) one last look? 🙏